### PR TITLE
Report progress for cache cleanup

### LINKED ISSUE #6024: Fully clean up caches and report progress
Instead of incrementally cleaning up caches, do full cleanups to ensure caches of users upgrading to the new version get cleaned up without requiring a large number of runs. In addition, for users who don't use the daemon we should provide a visual indication of what's going on at the end of their builds.

### PR DESCRIPTION
**Pull request by marcphilipp**
This PR makes cache cleanup non-incremental so when it is due a full cleanup is performed. Most cleanup operations are triggered on daemon shutdown. However, when the daemon is disabled, they are run after the build is finished. Previously, it looked like Gradle was unresponsive, especially for large caches. Now, progress is reported using the `ProgressLogger` API with status updates for each deleted/skipped cache entry. There's currently no percentage reported because in some cases we currently don't know the number of cache entries a priori (that could be changed, though). Another potential addition would be to count the number of deleted files and append it to the status.

Here's a screenshot of what it looks like:
screen shot 2018-07-19 at 15 37 37

Resolves #6024.

### CODE REVIEW

**Comment by marcphilipp** [comment]
> What happens when I kill the client while the cleanup is happening?

Nothing bad, it's just aborted and will start over next time.

**Comment by marcphilipp** [comment]
@oehme Any idea how to resolve these errors I'm now getting from integration tests on CI?

```
org.gradle.internal.service.ServiceCreationException: Cannot create service of type CacheFactory using GlobalScopeServices.createCacheFactory() as there is a problem with parameter #3 of type ProgressLoggerFactory.
	at org.gradle.internal.service.DefaultServiceRegistry$FactoryService.bind(DefaultServiceRegistry.java:750)
	at org.gradle.internal.service.DefaultServiceRegistry$SingletonService.prepare(DefaultServiceRegistry.java:639)
	at org.gradle.internal.service.DefaultServiceRegistry$SingletonService.getService(DefaultServiceRegistry.java:662)
	at org.gradle.internal.service.DefaultServiceRegistry$OwnServices.getService(DefaultServiceRegistry.java:440)
	at org.gradle.internal.service.DefaultServiceRegistry$CompositeServiceProvider.getService(DefaultServiceRegistry.java:923)
	at org.gradle.internal.service.DefaultServiceRegistry.find(DefaultServiceRegistry.java:1059)
	at org.gradle.internal.service.DefaultServiceRegistry.getService(DefaultServiceRegistry.java:297)
	at org.gradle.internal.service.DefaultServiceRegistry.find(DefaultServiceRegistry.java:291)
	at org.gradle.internal.service.DefaultServiceRegistry.get(DefaultServiceRegistry.java:281)
	at org.gradle.internal.service.DefaultServiceRegistry.get(DefaultServiceRegistry.java:276)
	at org.gradle.cache.internal.DefaultGeneratedGradleJarCacheIntegrationTest.$spock_initializeFields(DefaultGeneratedGradleJarCacheIntegrationTest.groovy:55)
Caused by: org.gradle.internal.service.ServiceCreationException: Cannot create service of type ProgressLoggerFactory using GlobalScopeServices.createProgressLoggerFactory() as required service of type OutputEventListener is not available.
	at org.gradle.internal.service.DefaultServiceRegistry$FactoryService.bind(DefaultServiceRegistry.java:758)
	at org.gradle.internal.service.DefaultServiceRegistry$SingletonService.prepare(DefaultServiceRegistry.java:639)
	at org.gradle.internal.service.DefaultServiceRegistry$SingletonService.getService(DefaultServiceRegistry.java:662)
	at org.gradle.internal.service.DefaultServiceRegistry$OwnServices.getService(DefaultServiceRegistry.java:440)
	at org.gradle.internal.service.DefaultServiceRegistry$CompositeServiceProvider.getService(DefaultServiceRegistry.java:923)
	at org.gradle.internal.service.DefaultServiceRegistry.find(DefaultServiceRegistry.java:1059)
	at org.gradle.internal.service.DefaultServiceRegistry$FactoryService.bind(DefaultServiceRegistry.java:748)
	... 10 more
```

**Comment by oehme** [comment]
Those exceptions are coming from compiler worker VMs which don't have progress logging capabilities (and probably also shouldn't do any cache cleanup).

**Comment by marcphilipp** [comment]
Yeah, they shouldn't. Can you point me to where their service registries are built?

**Comment by oehme** [comment]
I'm only aware of Scala doing this [here](https://github.com/gradle/gradle/blob/master/subprojects/language-scala/src/main/java/org/gradle/api/internal/tasks/scala/ZincScalaCompilerFactory.java#L165).

File: subprojects/core/src/integTest/groovy/org/gradle/cache/internal/DefaultGeneratedGradleJarCacheIntegrationTest.groovy [hunk 1]
```diff
@@ -49,6 +50,11 @@ class DefaultGeneratedGradleJarCacheIntegrationTest extends Specification {
 
     def DefaultServiceRegistry services = (DefaultServiceRegistry) ServiceRegistryBuilder.builder()
             .parent(NativeServicesTestFixture.getInstance())
+            .provider(new Object() {
```
Comment (oehme) [inline subprojects/core/src/integTest/groovy/org/gradle/cache/internal/DefaultGeneratedGradleJarCacheIntegrationTest.groovy]: Maybe we should have something like a NO_OP logging service registry that we can wire in these cases? That would avoid some repetition.

Comment (marcphilipp) [reply subprojects/core/src/integTest/groovy/org/gradle/cache/internal/DefaultGeneratedGradleJarCacheIntegrationTest.groovy]: Good idea! Done in 3b4b7114c36.

**Pull request merged**